capsules/src/at_commands_driver.rs

```diff
@@ -0,0 +1,237 @@
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    /// Setup callbacks.
+    ///
+    /// ### `subscribe_num`
+    ///
+    /// - `1`: Finished sending AT command.
+    /// - `2`: Received data from the radio.
+    fn subscribe(
+        &self,
+        subscribe_num: usize,
+        callback: Option<Callback>,
+        app_id: AppId,
+    ) -> ReturnCode {
+        match subscribe_num {
+            1 => self
+                .apps
+                .enter(app_id, |app, _| {
+                    app.tx_callback = callback;
+                    ReturnCode::SUCCESS
+                })
+                .unwrap_or_else(|err| err.into()),
+            2 => self
+                .apps
+                .enter(app_id, |app, _| {
+                    app.rx_callback = callback;
+                    ReturnCode::SUCCESS
+                })
+                .unwrap_or_else(|err| err.into()),
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    /// Initiate AT commands and other interactions with the radio.
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Driver check.
+    /// - `1`: Transmits a buffer passed via `allow`, up to the length
+    ///        passed in `arg1`.
+    /// - `2`: Receives into a buffer passed via `allow`, waiting for a `\n`.
+    fn command(&self, cmd_num: usize, arg1: usize, _: usize, appid: AppId) -> ReturnCode {
+        match cmd_num {
+            0 /* check if present */ => ReturnCode::SUCCESS,
+            1 => {
+                let len = arg1;
+                self.apps.enter(appid, |app, _| {
+                    app.tx_buffer.as_ref().map_or(ReturnCode::ENOMEM, |slice| {
+                        // Check that we aren't trying to send more than the
+                        // shared AppSlice.
+                        if len > slice.len() {
+                            ReturnCode::EINVAL
+                        } else {
+                            // Copy into our buffer and then TX.
+                            self.tx_buffer.take().map(|buffer| {
+                                for (i, c) in slice.as_ref()[0..len]
+                                    .iter()
+                                    .enumerate()
+                                {
+                                    if buffer.len() <= i {
+                                        break;
+                                    }
+                                    buffer[i] = *c;
+                                }
+
+                                self.tx_in_progress.set(appid);
+                                self.uart.transmit(buffer, len);
+                            });
+                            ReturnCode::SUCCESS
+                        }
+                    })
+                }).unwrap_or_else(|err| err.into())
+            }
+            2 => {
+                self.rx_buffer.take().map_or(ReturnCode::ENOMEM, |buffer| {
+                    self.rx_in_progress.set(appid);
```

Comment:
So this is a last one wins type situation. Are AT commands transactional enough that we can just deliver the received response to the app that made the associated request?

Reply:
I think that is exactly right. Given that this is low level control of a radio (like ble-serialization) I think assuming a single app (or well coordinated apps) is a reasonable starting point.